Firefox performance improved with highlight feature

jwallet opened the pull request:
I got the same issue as the issues mentionned here, I did not added my build index.js at this time in Options - Collapsable files / Delete files. So highlighter was trying to colorize 6 minified lines:
- x2  - 16,631 (these were actually taking 1sec each)
- x2 ~281,000
- x2  ~327,000 

the last 4 were causing the script to crash, they resolved after 9secs each, but even then firefox was not responding, just scrolling up and down, I had to kill the tab after some minutes to kill Prism from keep trying to highlight the whole lines

### firefox performance improvements
- autocollapse: run it first and return if successful to kill the propagation because it won't be rendered anyway, just like delete does.
- highligther: added promises when executing prism.highlightElement
- highlighter: lines over the arbitrary max length of 9999 will be considered as minified and ignored
- Prism: Updated prism form 1.15 to 1.20, I took the previous url from 1.15 to make sure I got the same addons and features from 1.15
- Updated syntax second try logic (related to `<add>` `<del>` tags)

### some tweak for diff selector
- Now logs the element in `console.error `
- `new Set` for selectors to make sure no duplicates are there
- Valid if the element from the selector is `aria-hidden` or `display:none`

close #348 
close #351 
close #266 

<!--
    Check those that apply, and delete the ones that don't
-->

-   [ ] I updated the CHANGELOG.md
-   [x] I tested the changes in this pull request myself
-   [ ] I added Automated Tests
-   [ ] I added an Option to enable / disable this feature
-   [ ] I updated the README.md, with pictures if necessary

---

jwallet committed:
updated main.js error while creating pr, dom not found

jwallet commented on src/main.js:
```diff
@@ -178,31 +178,44 @@ function codeReviewFeatures(config) {
+                manipulateSummary(this)
             }
+
+            if (this.matches(diffSelector)) {
```
looking for parent `diff` just to make sure it gets validated for `display:none` before running functions.

reyronald commented:
I can't reproduce this bug in the `dev` branch 😕 . I tried creating a PR while having all extension's options enabled, and with both the new UI and old UI experience.

Can you think of what might be different in your environment versus mine? I'd like to be able to reproduce this locally to be able to properly review the PR

jwallet commented:
using firefox and old experience, looks like the element target can be reached if the pr is small, when is bigger, some elements gets removed (diff tab) by bitbucket when in the commit tab. this does happen when developping, firefox has some memory management issues when reloading, it cannot remove unused memory correctly, resulting in unmanaged memory. and the webpage gets slow and the error pops, at this time, firefox takes around 6gb ram space. so maybe an issue when the selector is triggered at the same time it gets removed by the DOM. Slow computers might have the same issue.

I can't reproduce this bug as well, this morning, but I remember it was coming from `<section>` blocks. I will try to find the proper path to test it

jwallet committed:
- updated prism 1.15 to 1.20
- valid autocollapse first and return if successful to kill the propagation
- added promises to prism highlight
- lines over the arbitrary max length of 9999 will be considered as minified

jwallet commented on src/main.js:
```diff
@@ -178,31 +180,40 @@ function codeReviewFeatures(config) {
+        '#commit',
+        '#diff',
+    ]
     const diffSelector = 'section.bb-udiff'
```
added back the previous selector but kept validation hidden/visible line 197:202

jwallet commented on src/syntax-highlight/syntax-highlight.js:
```diff
@@ -39,7 +39,7 @@ export default function syntaxHighlight(diff, afterWordDiff) {
 
     const $diff = $(diff)
-    syntaxHighlightSourceCodeLines($diff)
+    // syntaxHighlightSourceCodeLines($diff)
```
i commented to know why we started the process and reapplied it later with afterWordDiff callback, it was another reason of performance issue, some lines lost their token/color syntax after the second highlight

jwallet replied:
2 tests are currently failing because i commented this line, so depends on the answer I will change the tests

jwallet committed:
removed debounce

jwallet committed:
Merge branch 'dev' into fix-missing-node-create-pr

jwallet committed:
code rev

jwallet committed:
syntaxe redo logic

jwallet commented on src/syntax-highlight/syntax-highlight.js:
```diff
@@ -39,10 +39,10 @@ export default function syntaxHighlight(diff, afterWordDiff) {
     const $diff = $(diff)
-    syntaxHighlightSourceCodeLines($diff)
+    syntaxHighlightSourceCodeLines($diff, 'pre.source:not([class*=language])')
 
```
first time we highlight we never have class language

jwallet commented on src/syntax-highlight/syntax-highlight.js:
```diff
@@ -39,10 +39,10 @@ export default function syntaxHighlight(diff, afterWordDiff) {
 
     afterWordDiff(() => {
-        syntaxHighlightSourceCodeLines($diff)
+        syntaxHighlightSourceCodeLines($diff, '.addition pre, .deletion pre')
```
second time, we will want to look for tags `pre:has(ins)` `pre:has(del)` but the word-diff block only have `<ins>` elements in the `addition` block, the `deletion` block will be re-rendered as well, so we have to re-highlight there too. so the selector is `.addition pre, .deletion pre` and anyway the mutation-observer will not ask to re-render a diff block that as no word-diff ins/del

jwallet replied:
![image](https://user-images.githubusercontent.com/23088305/81894154-1d1b4b00-957d-11ea-84dd-9e36fd7b85a8.png)

jwallet commented on src/syntax-highlight/syntax-highlight.js:
```diff
@@ -69,19 +69,38 @@ export default function syntaxHighlight(diff, afterWordDiff) {
+            new Promise((resolve, reject) => {
+                const { classList, firstChild, innerText } = preElement
+                if (
+                    firstChild.$$rbb_isSyntaxHighlighted &&
```
went bitbucket re-renders a line with word-diff it changes the content, it leaves the class and property there, so I check for childElementCount (tokens) to see if it's just plain text or it was highlighted

jwallet commented on src/syntax-highlight/syntax-highlight.js:
```diff
@@ -69,19 +69,36 @@ export default function syntaxHighlight(diff, afterWordDiff) {
+                    return
+                }
+
+                Prism.highlightElement(preElement)
```
Maybe we can use web worker, but Prism warn about using async highlight... web workers are slower and just limiting the text to a max length of 9999 is quite enough. Sure it will skip this line, but the ui does not freezes anymore

```jsx
Prims.highlightElement(element, true, () => {}); // async with web worker
```

jwallet committed:
test

jwallet committed:
Merge branch 'dev' into fix-missing-node-create-pr

jwallet committed:
fix build

jwallet committed:
autocollapse removed boolean result

jwallet committed:
missing selector for mutation observer

reyronald commented on src/syntax-highlight/syntax-highlight.js:
```diff
@@ -69,19 +72,36 @@ export default function syntaxHighlight(diff, afterWordDiff) {
+                const { classList, firstChild, innerText } = preElement
+
+                if (firstChild.$$rbb_isSyntaxHighlighted) {
+                    reject(new Error('Already highlighted'))
```
Do we have to reject with an Error here? Can't we just ignore this line, resolve, and skip to the next promise? Worried that we will polluting the console and stack with unhandled exceptions that are unactionable and don't really represent an error conceptually

jwallet replied:
no no need to add an error. you can just reject. sorry I lately left github for other stuff I do not think I will give more time to projects outside of those I am maintening.

reyronald replied:
No problem @jwallet ! Thanks a lot for all you've contributed so far! Greatly appreciated ♥

reyronald reviewed (commented):
Looks good to me! 

I'll be merging this by Monday, would like to wait to see if you can answer the question I left before then.

Thanks a lot for working on this as usual 🙏
